File: C++/199. Binary Tree Right Side View.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

/*
    Here we use Level Order Traversal
    in each level pickup the last element
*/
vector<int> rightSideView(TreeNode *root)
{
    vector<int> res;

    queue<TreeNode *> que;
    int level;

    if (root)
        que.push(root);

    while (!que.empty())
    {
        level = que.size();

        while (level--)
        {
            root = que.front();
            que.pop();

            if (level == 0)
                res.push_back(root->val);

            if (root->left)
                que.push(root->left);

            if (root->right)
                que.push(root->right);
        }
    }
    return res;
}

File: Java/199. Binary Tree Right Side View.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();

        if (root == null)
            return result;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        int level;

        while (!queue.isEmpty()) {
            level = queue.size();

            while (level-- > 0) {
                root = queue.poll();

                if (level == 0)
                    result.add(root.val);

                if (root.left != null)
                    queue.add(root.left);

                if (root.right != null)
                    queue.add(root.right);
            }
        }

        return result;
    }

    /* ignore-boiler-code */
    public class TreeNode {
        public int val;
        public TreeNode left;
        public TreeNode right;

        public TreeNode() {
        }

        public TreeNode(int val) {
            this.val = val;
        }

        public TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}